File: app/src/main/java/wooyun/esnb/view/RoundButton.java
package wooyun.esnb.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;
import android.util.AttributeSet;


public class RoundButton extends android.support.v7.widget.AppCompatButton {
    private final RectF roundRect = new RectF();
    private final Paint desPaint = new Paint();
    private final Paint srcPaint = new Paint();
    private float mRadius = 10;
    private boolean isChange = false;

    public RoundButton(Context context) {
        super(context);
        init();
    }

    public RoundButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public RoundButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void change(boolean isChange) {
        this.isChange = isChange;
        invalidate();
    }

    public void init() {
        desPaint.setAntiAlias(true);//设置抗锯齿
        desPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        srcPaint.setAntiAlias(true);
        float density = getResources().getDisplayMetrics().density;
        mRadius *= density;
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        int width = getWidth();
        int height = getHeight();
        roundRect.set(0, 0, width, height);
    }

    @Override
    public void draw(Canvas canvas) {
        //保存最原始的roundRect
        canvas.saveLayer(roundRect, srcPaint, Canvas.ALL_SAVE_FLAG);
        if (isChange) {
            //保存去掉头部2圆角的roundRect(实际就是保留底部的2个圆角)
            canvas.drawRect(roundRect.left, (roundRect.top + roundRect.bottom) / 2, roundRect.right, roundRect.bottom, srcPaint);
            //保存去掉底部2圆角的roundRect(实际就是保留顶部的2个圆角)
            canvas.drawRect(roundRect.left, roundRect.top, roundRect.right, roundRect.bottom / 2, srcPaint);
        }
        //保存掉头部2圆角的roundRect
        canvas.drawRoundRect(roundRect, mRadius, mRadius, srcPaint);
        //保存叠加后的内容
        canvas.saveLayer(roundRect, desPaint, Canvas.ALL_SAVE_FLAG);
        super.draw(canvas);
        //清空所有的图像矩阵修改状态
        canvas.restore();
    }
}

File: app/src/main/java/wooyun/esnb/util/RoundButton.kt
package wooyun.esnb.util

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.os.Build
import android.support.annotation.RequiresApi
import android.view.View

class RoundButton(context: Context) : View(context) {
    private val paint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var width: Float? = null
    private var hight: Float? = null


    fun setKey(width: Float, hight: Float) {
        this.width = width
        this.hight = hight
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        val left = left.toFloat()
        val right = right.toFloat()
        val top = top.toFloat()
        val bottom = bottom.toFloat()
        canvas!!.drawRoundRect(left, top, right, bottom, width!!, hight!!, paint)
    }
}